### PR TITLE
feat(agents-api): Adaptive context (via recursive summarization)

### LINKED ISSUE #328: Feat: Implement new recursive summarization logic
image

### PR DESCRIPTION
**Pull request by creatorrr**
Closes #328 

----

| :rocket: | This description was created by [Ellipsis](https://www.ellipsis.dev) for commit 0ddc2713d217fa0321e412ade28182ebe3b9f586  | 
|--------|--------|

### Summary:
This PR enhances the `agents-api` by improving session management, message summarization, introducing new features, and updating the OpenAPI model.

**Key points**:
- Enhances `agents-api` by improving session management, message summarization, and introducing new features like `TruncationWorkflow`, along with updates to the OpenAPI model and session handling logic.
- Refactored `summarize_messages` in `summarization.py` for efficiency.
- Introduced `token_budget` and `context_overflow` for dynamic session management.
- Updated OpenAPI model and session handling logic.
- Fixed various issues across modules for improved functionality.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

File: agents-api/agents_api/rec_sum/entities.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+@retry(stop=stop_after_attempt(2))
+async def get_entities(
+    chat_session,
+    model="gpt-4-turbo",
+    stop=["</ct:entities"],
+    temperature=0.7,
+    **kwargs,
+):
+    assert len(chat_session) > 2, "Session is too short"
+
```
Comment (creatorrr) [inline agents-api/agents_api/rec_sum/entities.py:60]: So this produces `{"role": "system", "name": "entities", "content": ...}` messages.

But this part needs some logic for handling a pre-existing entities message. If one exists then the new one will replace the old one. Maybe this needs to happen at the part where the resulting messages are saved with a relation.

File: agents-api/agents_api/activities/summarization.py [hunk 2]
```diff
@@ -148,31 +152,76 @@ async def run_prompt(
...
+    add_entries_query(
+        Entry(
+            session_id=session_id,
+            source="summarizer",
+            role="system",
+            name="entities",
+            content=entities["content"],
+            timestamp=entries[0]["timestamp"] + ts_delta,
+        )
     )
```
Comment (creatorrr) [inline agents-api/agents_api/activities/summarization.py:212]: (See comment below about the entries query)

> So this produces {"role": "system", "name": "entities", "content": ...} messages.

> But this part needs some logic for handling a pre-existing entities message. If one exists then the new one will replace the old one. Maybe this needs to happen at the part where the resulting messages are saved with a relation.

**Comment by hbrooks** [comment]
Hey @creatorrr, founder @ellipsis-dev here. You can disable code reviews from certain branch names, such as into `dev` or from `f/*` by adding a config file: https://docs.ellipsis.dev/config#ignore-certain-basehead-branches.

**Comment by creatorrr** [comment]
> Hey @creatorrr, founder @ ellipsis-dev here. You can disable code reviews from certain branch names, such as into `dev` or from `f/*` by adding a config file: https://docs.ellipsis.dev/config#ignore-certain-basehead-branches.

Thanks for the pro tip @hbrooks :)

File: agents-api/agents_api/common/protocol/entries.py [hunk 3]
```diff
@@ -49,6 +49,7 @@ def token_count(self) -> int:
                     # TODO: how to calc token count for images?
                     if isinstance(part, ChatMLTextContentPart):
                         text += part.text
+                content_length = len(text)
```
[the remark this answers is not in the document]
Comment (creatorrr) [reply agents-api/agents_api/common/protocol/entries.py:52]: @whiterabbit1983 ☝️

**Comment by creatorrr** [comment]
We need a flag for adaptive context, maybe an environment variable like `EXPERIMENTAL_ADAPTIVE_CONTEXT_ON`

**Comment by creatorrr** [comment]
and if content is an array, the text parts should be concatenated with a single space and the images need to be replaced with the string `<UPLOADED IMAGE>`

I think that should work without changing the prompt but the tricky thing will be to test it and also still keep the image

File: agents-api/agents_api/activities/summarization.py [hunk 4]
```diff
@@ -160,28 +164,56 @@ async def run_prompt(
 @activity.defn
 async def summarization(session_id: str) -> None:
     session_id = UUID(session_id)
```
Comment (creatorrr) [inline agents-api/agents_api/activities/summarization.py:166]: [only in adaptive mode] add some logic to handle messages that contain images. If `message.image_url.detail != "low"` then produce a new entry that "summarizes" that entry by setting `detail` to `"low"`

Comment (creatorrr) [reply agents-api/agents_api/activities/summarization.py:166]: apply this first, see if the token threshold is still being breached and then apply the summarize step. we should do this before the trim operation too -- only apply the `trim` operation if the summarize and entities step result is still greater than the token threshold

**Pull request merged**